### ISSUE TITLE
Avoid Device.BeginInvokeOnMainThread() in libraries

### ISSUE THREAD

**Issue by KPixel** [billvenhaus/ReorderableCollectionView.Maui#1]
Hi,

First, this library is really awesome. Well done.

I did notice that you are [using Device.BeginInvokeOnMainThread() in a bunch of places](https://github.com/billvenhaus/ReorderableCollectionView.Maui/search?q=Device.BeginInvokeOnMainThread). Please, avoid that. It makes the library impossible to use in apps that have more than one window (which can happen on Windows, and is becoming more common on Android/iOS).

The correct approach is to capture the dispatcher from the UI. Here is a practical example: [MarshalingObservableCollection in MAUI](https://github.com/dotnet/maui/blob/1cd6ed9e1736f074aa3340c48e5a7913bd77c841/src/Controls/src/Core/Items/MarshalingObservableCollection.cs).

**Comment by billvenhaus** [comment]
Thanks!

Most occurrences you'll find are in the "ItemSources" folders. Those files were almost all cloned from the main MAUI repo. I definitely did not add any `Device.BeginInvokeOnMainThread` calls there. I only wished to make the smallest changes necessary to those files for reordering support. You can see where the file was sourced from at the top in case you want to compare the changes I made.

There's actually only 1 call I added for `Device.BeginInvokeOnMainThread`. That's in [ReorderableCollectionView.cs](https://github.com/billvenhaus/ReorderableCollectionView.Maui/blob/master/src/ReorderableCollectionView.Maui/ReorderableCollectionView.cs) & it is only used for` VariableSpanGridItemsLayout.`

**Comment by KPixel** [comment]
Xamarin.Forms and MAUI do have this problem in some places.

Regarding your code:

https://github.com/billvenhaus/ReorderableCollectionView.Maui/blob/851da3e73cc8934ad283dcd5c792414e4e018f28/src/ReorderableCollectionView.Maui/ReorderableCollectionView.cs#L37-L40

All views have a `Dispatcher` property. So, you can replace it with `Dispatcher.DispatchIfRequired()`.

It will also fix an (unrelated) bug where `InvalidateMeasureNonVirtual()` would not be called if `!IsInvokeRequired`.

**Comment by billvenhaus** [comment]
> It will also fix an (unrelated) bug where `InvalidateMeasureNonVirtual()` would not be called if `!IsInvokeRequired`.

Good catch!  I missed that. I'll fix that in the next release. I'll create a new build when MAUI Preview 11 is published.

I'll also update it to use the Dispatcher there instead. Although it looks like `Dispatcher.DispatchIfRequired()` is an extension method in an internal class.
https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/DispatcherExtensions.cs

**Comment by KPixel** [comment]
Oh, I didn't notice that that class was internal. IMO, it should be public since it is a very common pattern.

For now, you can create your own (simpler):

```C#
		public static void DispatchIfRequired(this IDispatcher dispatcher, Action action)
		{
			if (dispatcher.IsDispatchRequired)
			{
				dispatcher.Dispatch(action);
			}
			else
			{
				action();
			}
		}
```
